nengo_spa/algebras/tvtb_algebra.py

```diff
@@ -0,0 +1,206 @@
+import nengo
+import numpy as np
+
+from nengo_spa.algebras.base import AbstractAlgebra, ElementSidedness
+from nengo_spa.networks.tvtb import TVTB
+
+
+class TvtbAlgebra(AbstractAlgebra):
+    r"""Transposed Vector-derived Transformation Binding (TVTB) algebra.
+
+    TVTB uses elementwise addition for superposition. The binding operation
+    :math:`\mathcal{B}(x, y)` is defined as
+
+    .. math::
+
+       \mathcal{B}(x, y) := V_y^T x = \left[\begin{array}{ccc}
+           V_y'^T &      0 &      0 \\
+              0   & V_y'^T &      0 \\
+              0   &      0 & V_y'^T
+           \end{array}\right] x
+
+    with
+
+    .. math::
+
+       V_y' = d^{\frac{1}{4}} \left[\begin{array}{cccc}
+           y_1            & y_2            & \dots  & y_{d'}  \\
+           y_{d' + 1}     & y_{d' + 2}     & \dots  & y_{2d'} \\
+           \vdots         & \vdots         & \ddots & \vdots  \\
+           y_{d - d' + 1} & y_{d - d' + 2} & \dots  & y_d
+       \end{array}\right]
+
+    and
+
+    .. math:: d'^2 = d.
+
+    The approximate inverse :math:`y^+` for :math:`y` is permuting the elements
+    such that :math:`V_{y^+} = V_y^T`.
+
+    Note that TVTB requires the vector dimensionality to be square.
+
+    The TVTB binding operation is neither associative nor commutative.
+    In contrast to VTB, however, TVTB has two-sided identities and inverses.
+    Other properties are equivalent to VTB.
+
+    .. seealso::
+        `.VtbAlgebra`
+    """
+
+    _instance = None
+
+    def __new__(cls):
+        if type(cls._instance) is not cls:
+            cls._instance = super(TvtbAlgebra, cls).__new__(cls)
+        return cls._instance
+
+    def is_valid_dimensionality(self, d):
+        """Checks whether *d* is a valid vector dimensionality.
+
+        For TVTB all square numbers are valid dimensionalities.
+
+        Parameters
+        ----------
+        d : int
+            Dimensionality
+
+        Returns
+        -------
+        bool
+            *True*, if *d* is a valid vector dimensionality for the use with
+            the algebra.
+        """
+        if d < 1:
+            return False
+        sub_d = np.sqrt(d)
+        return sub_d * sub_d == d
+
+    def _get_sub_d(self, d):
+        sub_d = int(np.sqrt(d))
+        if sub_d * sub_d != d:
+            raise ValueError("Vector dimensionality must be a square number.")
+        return sub_d
+
+    def make_unitary(self, v):
```

Comment:
Is there a name or explanation for this algorithm?

Reply:
Maybe ... probably? It's generating an orthogonal matrix, same algorithm as the original VTB.

Reply:
Is it the [Gram-Schmidt process](https://en.wikipedia.org/wiki/Gram%E2%80%93Schmidt_process)?

Reply:
I think I even left the same comment when reviewing the original VTB...

Once again, upon further consideration, I think I can trust the name `make_unitary` even if I don't completely grasp the internals.